core/DbHelper.php

```diff
@@ -204,6 +204,62 @@ public static function tableHasIndex($table, $indexName)
         return !empty($result);
     }
 
+    /**
+     * Returns the default database charset to use
+     *
+     * Returns utf8mb4 if supported, with fallback to utf8
+     *
+     * @return string
+     * @throws Tracker\Db\DbException
+     */
+    public static function getDefaultCharset()
+    {
+        $result = Db::get()->fetchRow("SHOW CHARACTER SET LIKE 'utf8mb4'");
+
+        if (empty($result)) {
+            return 'utf8'; // charset not available
+        }
+
+        $result = Db::get()->fetchRow("SHOW VARIABLES LIKE 'character_set_database'");
+
+        if (!empty($result) && $result['Value'] === 'utf8mb4') {
+            return 'utf8mb4'; // database has utf8mb4 charset, so assume it can be used
+        }
+
+        $result = Db::get()->fetchRow("SHOW VARIABLES LIKE 'innodb_file_per_table'");
+
+        if (empty($result) || $result['Value'] !== 'ON') {
+            return 'utf8'; // innodb_file_per_table is required for utf8mb4
+        }
+
+        return 'utf8mb4';
```

Comment:
Similar to https://docs.nextcloud.com/server/18/admin_manual/configuration_database/mysql_4byte_support.html do we also need to check if ` show variables like 'innodb_file_format' == 'Antelope '` and if so prefer utf8?

Reply:
Actually not, as we currently do not require `large_prefix_index`, which doesn't work with `Antelope`. If we decide to require `large_prefix_index`, to make it possible to have key length greater than 255, we also need to require `Barracuda`